assets/js/components/ResetButton.js

```diff
@@ -0,0 +1,103 @@
+/**
+ * ResetButton component.
+ *
+ * Site Kit by Google, Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { Fragment, useState, useEffect, useCallback } from '@wordpress/element';
+import { addQueryArgs } from '@wordpress/url';
+
+/**
+ * Internal dependencies
+ */
+import Data from 'googlesitekit-data';
+import {
+	clearWebStorage,
+	sanitizeHTML,
+} from '../util';
+import Dialog from './dialog';
+import Link from './Link';
+import Modal from './modal';
+import { STORE_NAME as CORE_SITE } from '../googlesitekit/datastore/site/constants';
+const { useSelect, useDispatch } = Data;
+
+function ResetButton( { children } ) {
+	const splashURL = useSelect( ( select ) => select( CORE_SITE ).getAdminURL( 'googlesitekit-splash' ) );
+
+	const [ dialogActive, setDialogActive ] = useState( false );
+	const postResetURL = addQueryArgs( splashURL, { notification: 'reset_success' } );
+
+	const { reset: resetCoreSite } = useDispatch( CORE_SITE );
+
+	const handleCloseModal = useCallback( ( event ) => {
+		if ( 27 === event.keyCode ) {
+			setDialogActive( false );
+		}
+	}, [ dialogActive ] );
+
+	useEffect( () => {
+		global.addEventListener( 'keyup', handleCloseModal, false );
+
+		return () => global.removeEventListener( 'keyup', handleCloseModal );
```

Comment:
> **Note**
> 
> React guarantees that setState function identity is stable and won’t change on re-renders. This is why it’s safe to omit from the useEffect or useCallback dependency list.

According to [this](https://reactjs.org/docs/hooks-reference.html#usestate), `setDialogActive` will always be the same as `setState` is always same. This means the `useCallback` value (`handleCloseModal`) will never change. In turn, the `useEffect` will never run except component mount and unmount. 

What I think should be instead here is this (ignore comments and console logs for now):
```js
const handleCloseModal = useCallback( ( event ) => {
	if ( 27 === event.keyCode ) {
		// Only close the modal if the "Escape" key is used.
		setDialogActive( false );
	}
	// eslint-disable-next-line no-console
	console.log( 'event callback called' );
// as it depends on nothing, an empty dependency array
}, [ ] );

useEffect( () => {
	if ( dialogActive ) {
                // only when the dialog is shown, add the event listener
		// eslint-disable-next-line no-console
		console.log( 'adding listener' );
		global.addEventListener( 'keyup', handleCloseModal, false );
	}

	return () => {
		global.removeEventListener( 'keyup', handleCloseModal );
		// eslint-disable-next-line no-console
		console.log( 'removing listener' );
	};
// will be run when the dialog is shown or hidden
}, [ dialogActive ] );
```

In this case, the listener is removed when the dialog is closed and added when the dialog is opened. What do you think 🤔 ?

Reply:
Yeah, I actually thought the `setState` function wouldn't change, but had it in there for completeness. But I suspected `useCallback` with no arguments will work.

I think you can remove the `[]` argument altogether, but otherwise looks good.

-----

Regarding having `useEffect` update when `dialogActive` changes: that's fine, but your proposal isn't quite right because technically every time it changes and `dialogActive` is `true`, _another_ event handler will be attached as I understand it.

I forget if attaching the same handler to an event twice will actually cause it to be run twice, but we don't need that.

(In your suggestion, the event listener will persist until the component unmounts.)

Basically, we want:

1. mount the component
2. define the callback
3. attach the `handleCloseModal` handlers

and then everything just stays as-is until we unmount, when the handler should be unmounted.

But I guess removing the listener when the dialog disappears is actually a good call, so let's do that:

```js
const handleCloseModal = useCallback( ( event ) => {
	if ( 27 === event.keyCode ) {
		// Only close the modal if the "Escape" key is used.
		setDialogActive( false );
	}
} );

useEffect( () => {
	if ( dialogActive ) {
		global.addEventListener( 'keyup', handleCloseModal, false );

        return () => {
            global.removeEventListener( 'keyup', handleCloseModal );
        };
	} else {
        // Remove the event listener when the dialog is removed; there's no need
        // to have it attached when it won't be used.
        global.removeEventListener( 'keyup', handleCloseModal );
    }
}, [ dialogActive ] );
```